src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java

```diff
@@ -0,0 +1,593 @@
+     */
+    public void setProtectedForwardPath(HaFlowPath forwardPath) {
+        if (forwardPath != null) {
+            validateForwardPath(forwardPath);
+            if (!hasPath(forwardPath)) {
+                addPaths(forwardPath);
+            }
+            setProtectedForwardPathId(forwardPath.getHaPathId());
+        } else {
+            setProtectedForwardPathId(null);
+        }
+    }
+
+    /**
+     * Get the reverse path.
+     */
+    public HaFlowPath getReversePath() {
+        if (getReversePathId() == null) {
+            return null;
+        }
+        return getPath(getReversePathId()).orElse(null);
+    }
+
+
+    /**
+     * Add a path and set it as the reverse path.
+     */
+    public void setReversePath(HaFlowPath reversePath) {
+        if (reversePath != null) {
+            if (!hasPath(reversePath)) {
+                validateReversePath(reversePath);
+                addPaths(reversePath);
+            }
+            setReversePathId(reversePath.getHaPathId());
+        } else {
+            setReversePathId(null);
+        }
+    }
+
+    /**
+     * Get the protected reverse path.
+     */
+    public HaFlowPath getProtectedReversePath() {
+        if (getProtectedReversePathId() == null) {
+            return null;
+        }
+        return getPath(getProtectedReversePathId()).orElse(null);
+    }
+
+    /**
+     * Add a path and set it as the protected reverse.
+     */
+    public void setProtectedReversePath(HaFlowPath reversePath) {
+        if (reversePath != null) {
+            if (!hasPath(reversePath)) {
+                validateReversePath(reversePath);
+                addPaths(reversePath);
+            }
+            setProtectedReversePathId(reversePath.getHaPathId());
+        } else {
+            setProtectedReversePathId(null);
+        }
+    }
+
+    private void validateEndpoints(HaFlowPath path) {
+        if (!path.getSharedSwitchId().equals(getSharedEndpoint().switchId)) {
+            throw new IllegalArgumentException(format("HA-path %s has shared endpoint switch id %s, but %s expected",
+                    path.getHaPathId(), path.getSharedSwitchId(), getSharedEndpoint().switchId));
+        }
+        Set<SwitchId> subFlowSwitchIds = getSubFlows().stream()
+                .map(HaSubFlow::getEndpointSwitchId)
+                .collect(Collectors.toSet());
+        Set<SwitchId> pathSwitchIds = path.getSubFlowSwitchIds();
+
+        if (!subFlowSwitchIds.equals(pathSwitchIds)) {
+            throw new IllegalArgumentException(format("HA-path %s must have %s switch endpoints, but has following "
+                            + "endpoints: %s",
+                    path.getHaPathId(), subFlowSwitchIds, pathSwitchIds));
+        }
+    }
+
+    private void validateForwardPath(HaFlowPath haPath) {
+        validateEndpoints(haPath);
+        if (!haPath.isForward()) {
+            throw new IllegalArgumentException(format("Cookie %s of HA-path %s must have forward direction",
```

Comment:
To me, in this case, "direction" is a countable noun (meaning a route or a path, right?), so, _**a**_ forward direction

Reply:
I imagine this sort of "direction" as an arrow. There one direction: forward, to the right, there is another direction, reverse, to the left. So, two directions in total. I can count them -> countable.
Direction could be uncountable when it means an abstract idea. It's something like when we speak about sea turtles and their ability to know where they are and where to go without a compass, or they have a sense of direction (no specific arrow here, here it means an ability to know where to go and don't get lost on the way). 
For switches, I think it is the first case.

Reply:
fixed